Support for new C*2.2 date and time types NODEJS-118

jorgebay opened the pull request.

tolbertam commented on lib/types/local-date.js:
```diff
@@ -0,0 +1,200 @@
+      return self.value.getUTCFullYear();
+    }},
+    'month': { enumerable: true, get: function () {
+      return self.value.getMonth() + 1;
```
I think these should get getUTCMonth and getUTCDay as given Date.UTC is used to create the date.  The tests in encoder-tests.js fail if local time is offset enough from UTC.

jorgebay replied:
Nice catch: I will fix it

tolbertam commented on lib/types/local-date.js:
```diff
@@ -0,0 +1,200 @@
+/**
+ * Creates a new instance of LocalDate using the current year, month and day from the system clock in the default time-zone.
+ */
+LocalDate.now = function () {
```
I think we should either assume that UTC is always used, or always use the Local time definition for defining date, otherwise this can be confusing.

tolbertam replied:
We could also consider getting rid of the now, utcNow, and fromDate methods altogether since it is easy enough to provide year, month, day itself.

jorgebay replied:
From the user perspective, `LocalDate` should always use local clock values (and does use local clock). The only case where utc is used / referenced is `utcNow()`, but it can be considered as a utility method. I took the concepts from [LocalDate](http://joda-time.sourceforge.net/apidocs/org/joda/time/LocalDate.html#LocalDate()).

Internally it uses UTC values to use the built-in calendar functionality, but the user doesn't have to know / understand that.

tolbertam replied:
Ah yes, now I'm in sync.  I was getting wrapped up in the internal implementation, but it turns out to not matter if you store it in UTC and look at it in UTC.   If you look at LocalDate.value with your local timezone, there's a chance it could appear off, but I think that is a consideration one needs to take into account when using value.

tolbertam commented on lib/types/local-date.js:
```diff
@@ -0,0 +1,200 @@
+    multiplier = -1;
+  }
+  var parts = value.split('-');
+  return new LocalDate(multiplier * parseInt(parts[0], 10), parseInt(parts[1], 10), parseInt(parts[2], 10));
```
Do you think it would be worth making the fromString implementations of LocalDate and LocalTime have more stricter validation of integer parsing?  parseInt will simply ignore non-numerical values, so you can create some unexpected behavior if non-complete numeric values are provided.

tolbertam commented on lib/types/local-date.js:
```diff
@@ -0,0 +1,200 @@
+ * @param {Number} day Between 1 and the number of days in the given month of the given year.
+ * @constructor
+ */
+function LocalDate(year, month, day) {
```
Since LocalDate and LocalTime are currently timezone independent, and the concept of 'local' implies local timezone (at least to me), should we consider renaming these?

jorgebay replied:
The names are based on [Java's LocalDate and LocalTime] (https://docs.oracle.com/javase/8/docs/api/java/time/LocalDate.html). 

I think it is meant to be local from any user perspective: "She was born on October 1st, 1980" (no matter where she lived or where she lives now, her birthday is going to remain the same).

tolbertam replied:
ah excellent.  Thanks for the explanation, since it is being patterned of an already existing implementation I would say this is fine then.  I was initially thrown off by Local == 'Local Timezone' but your explanation has me on the right page now.

tolbertam commented on lib/types/local-time.js:
```diff
@@ -0,0 +1,306 @@
+  if (typeof nanoseconds !== 'number') {
+    //generate
+    nanoseconds = nanoGenerator++;
+    if (nanoGenerator === 1000000) {
```
I think if we can't guarantee a monotonic value we should just use '0' instead if nanoseconds if not specified.

jorgebay replied:
Agree! Also, now that I think about it, there is no point in generating it for the user (times are not used for timestamps or anything like it), default to `0` would make more sense.

tolbertam commented on test/unit/basic-tests.js:
```diff
@@ -146,6 +146,131 @@ describe('types', function () {
+        values.forEach(function (item) {
+          var val = new LocalTime(Long.fromString(item[0]));
+          var parts = item[2];
+          console.log(item[1]);
```
extra console.log not needed.

jorgebay commented:
Added validation for generated date (via `fromString()` or any other).
Fixed `month` and `day`.
Fixed representation for first century dates.
Removed nanoseconds generation.
Throw `TypeError` for date and time encoder errors.

tolbertam commented:
Changes look great! :+1:
